src/Nethermind/Nethermind.Synchronization/StateSync/StateSyncDispatcher.cs

```diff
@@ -14,42 +14,148 @@
 using Nethermind.Synchronization.Peers;
+using Nethermind.Trie;
 
 namespace Nethermind.Synchronization.StateSync
 {
     public class StateSyncDispatcher : SyncDispatcher<StateSyncBatch>
     {
-        public StateSyncDispatcher(ISyncFeed<StateSyncBatch> syncFeed, ISyncPeerPool syncPeerPool, IPeerAllocationStrategyFactory<StateSyncBatch> peerAllocationStrategy, ILogManager logManager)
+        //private string _trace;
+
+        private readonly bool _snapSyncEnabled;
+
+        public StateSyncDispatcher(ISyncFeed<StateSyncBatch> syncFeed, ISyncPeerPool syncPeerPool, IPeerAllocationStrategyFactory<StateSyncBatch> peerAllocationStrategy, ISyncConfig syncConfig, ILogManager logManager)
             : base(syncFeed, syncPeerPool, peerAllocationStrategy, logManager)
         {
+            _snapSyncEnabled = syncConfig.SnapSync;
         }
 
-        protected override async Task Dispatch(PeerInfo peerInfo, StateSyncBatch request, CancellationToken cancellationToken)
+        protected override async Task Dispatch(PeerInfo peerInfo, StateSyncBatch batch, CancellationToken cancellationToken)
         {
+            if (batch == null || batch.RequestedNodes == null || batch.RequestedNodes.Length == 0)
+            {
+                return;
+            }
+
+            //string printByteArray(byte[] bytes)
+            //{
+            //    return string.Join(null, bytes.Select(b => b.ToString("X")));
+            //}
+            //_trace += String.Join(Environment.NewLine, batch.RequestedNodes.Select(n => n.Level + "|" + n.NodeDataType + "|" + printByteArray(n.AccountPathNibbles) + "|" + printByteArray(n.PathNibbles)))
+            //    + Environment.NewLine;
+
             ISyncPeer peer = peerInfo.SyncPeer;
-            var getNodeDataTask = peer.GetNodeData(request.RequestedNodes.Select(n => n.Hash).ToArray(), cancellationToken);
-            await getNodeDataTask.ContinueWith(
+
+            Task<byte[][]> task = null;
+
+            if (_snapSyncEnabled)
+            {
+                if (peer.TryGetSatelliteProtocol<ISnapSyncPeer>("snap", out var handler))
+                {
+                    if (batch.NodeDataType == NodeDataType.Code)
+                    {
+                        var a = batch.RequestedNodes.Select(n => n.Hash).ToArray();
+                        Logger.Warn($"GETBYTECODES count:{a.Length}");
+                        task = handler.GetByteCodes(a, cancellationToken);
+                    }
+                    else
+                    {
+                        GetTrieNodesRequest request = GetRequest(batch);
+
+                        Logger.Warn($"GETTRIENODES count:{request.AccountAndStoragePaths.Length}");
+
+                        task = handler.GetTrieNodes(request, cancellationToken);
+                    }
+                }
+            }
+
+            if (task is null)
+            {
+                var a = batch.RequestedNodes.Select(n => n.Hash).ToArray();
```

Comment:
do we have to ToArray here?

Reply:
Related to:
https://github.com/NethermindEth/nethermind/issues/4377